keps/sig-apps/20190412-pdb-for-scale-subresource.md

```diff
@@ -0,0 +1,84 @@
+
+The scale subresource allow any resource to specify its desired number of replicas and a generic way to look up this information. This document lays out a plan to use the scale subresource to allow setting PDBs for any resource implementing the scale subresource.
+
+## Motivation
+
+PDBs are an important tool to control the number of voluntary disruptions for workloads on Kubernetes. As more users start deploying custom controllers/operators based on CRDs, it is inconvenient that PDBs are not available. The scale subresource is already in use by the autoscaler and we should extend its use to PDBs.
+
+### Goals
+
+- Implement support for scale subresources in PDBs
+- Avoid major performance impact from the change
+
+## Proposal
+
+### Implementation Details/Notes/Constraints
+
+In the current implementation, the PDB controller identifies the workload controller by going through all the pods covered by the PDB, and for each pod, check in sequence for each of the four workload controllers. The PDB controller then looks up the desired number of replicas from the identified controller. If the set of pods covered by the PDB identfies more than one controller, that is considered an error. 
+
+The plan is to keep this approach, but check for the scale subresource if neither of the kubernetes workload controllers match. Each of the kubernetes workload controllers actually implement the scale subresource, so it is possible to only use that one. But since shared informers can not be used with the scale subresource, we want to rely on the informers for checking the kubernetes workload controllers, and only try the scale subresource if none of the them match.
+
+As mentioned above, the PDB controller will look up the controller for every pod covered by the PDB. In almost all scenarios the controller will be the same for all pods and therefore the desired number of replicas will be the same. This is not a concern for the kubernetes workload controllers since the controller uses shared informers. When using the scale subresource, this would lead to unnecessary calls to the apiserver. We will implement a per reconcile loop cache, so the PDB controller only needs to look up each controller once per reconcile loop.
+
+### Risks and Mitigations
+
+The major risk with this change is the additional load on the apiserver since we can't use shared informers for scale subresources. As described above, we plan to mitigate this by only checking the scale subresource when neither of the kubernetes workload controllers is a match and use a cache to make sure we only need to look up each controller once per reconcile loop.
```

Comment:
Is it a reasonable assumption to avoid queyring the api server for pods w/o an owner ref?

Reply:
Pods without a controller ref needn't be looked up at all, we know there's not a controller managing them. At least not a correctly written controller.

Reply:
My thoughts, worth making it more explicit in the proposal.

Reply:
I have tried to clarify this in the proposal. The PDB controller will only query the apiserver when there is a controller ref for a pod, but the kind doesn't match any of the kubernetes workload controllers.